src/rust/engine/src/nodes.rs

```diff
@@ -710,25 +711,46 @@ impl Node for SelectTransitive {
       .then(move |dep_product_res| {
         match dep_product_res {
           Ok(dep_product) => {
-            // The product and its dependency list are available: project them.
-            let deps =
-              future::join_all(
-                externs::project_multi(&dep_product, &self.selector.field).iter()
-                  .map(|dep_subject| self.get_dep(&context, &dep_subject))
-                  .collect::<Vec<_>>()
-              );
-            deps
-              .then(move |dep_values_res| {
-                // Finally, store the resulting values.
-                match dep_values_res {
-                  Ok(dep_values) => {
-                    Ok(self.store(&dep_product, dep_values.iter().collect()))
-                  },
-                  Err(failure) =>
-                    Err(context.was_required(failure)),
+            let addresses = externs::project_multi(&dep_product, &self.selector.field);
+
+            let init = TransitiveExpansion {
+              todo: addresses,
+              completed: HashSet::default(),
+              outputs: OrderMap::default()
+            };
+
+            loop_fn(init, move |mut expansion| {
+              let round = future::join_all({
+                 expansion.todo.drain(..)
+                   .map(|address| self.expand_transitive(&context, &address))
+                   .collect::<Vec<_>>()
+              });
+
+            round
+              .map(move |finished_items| {
+                for (address, hydrated_target, more_deps) in finished_items.into_iter() {
+                  let address_key = externs::key_for(&address);
+                  expansion.completed.insert(address_key);
+
+                  expansion.outputs.insert(address_key, hydrated_target);
+
+                  let completed = &mut expansion.completed;
+                  expansion.todo.extend(more_deps.into_iter()
+                    .filter(|dep| !completed.contains(&externs::key_for(dep)))
```

Comment:
This computes the `Key` a third time. Maybe have the `todo` list be a tuple of the `Key` and the `Value` for a particular item?